### PR TITLE
Add Rect attribute kwargs for Surface.blit 

### PR DESCRIPTION
**Pull request by novialriptide**
Fixes: https://github.com/pygame/pygame/issues/3202

### CODE REVIEW

File: buildconfig/stubs/pygame/surface.pyi [hunk 1]
```diff
@@ -52,6 +52,23 @@ class Surface:
...
+        topleft: Union[Coordinate, RectValue, None] = None,
+        topright: Union[Coordinate, RectValue, None] = None,
+        bottomleft: Union[Coordinate, RectValue, None] = None,
+        bottomright: Union[Coordinate, RectValue, None] = None,
+        midleft: Union[Coordinate, RectValue, None] = None,
+        midright: Union[Coordinate, RectValue, None] = None,
+        midtop: Union[Coordinate, RectValue, None] = None,
+        midbottom: Union[Coordinate, RectValue, None] = None,
+        center: Union[Coordinate, RectValue, None] = None,
+        area: Optional[RectValue] = None,
```
Comment (cclauss) [inline buildconfig/stubs/pygame/surface.pyi:68]: OPTIONAL: Just FYI, mypy will not complain on function parameters if you do:
```suggestion
        topleft: Union[Coordinate, RectValue] = None,
        topright: Union[Coordinate, RectValue] = None,
        bottomleft: Union[Coordinate, RectValue] = None,
        bottomright: Union[Coordinate, RectValue] = None,
        midleft: Union[Coordinate, RectValue] = None,
        midright: Union[Coordinate, RectValue] = None,
        midtop: Union[Coordinate, RectValue] = None,
        midbottom: Union[Coordinate, RectValue] = None,
        center: Union[Coordinate, RectValue] = None,
        area: RectValue = None,
```

**Review by ankith26 — COMMENTED** [review]
First of all, thanks for the PR! This is indeed a useful feature! 🥳 😎 

As MyreMylar expressed in the linked issue how this could be a new function, I quite like that suggestion myself too

I'm not against the current API proposed... but I think we can investigate a few other ways of doing this (and hopefully get more feedback from more people on this since this seems like an extensive change)

I have an idea in mind, I will elaborate on it in the linked issue.

Additionally, this PR needs more extensive unit testing (this is after we can finalise the API first)

**Comment by novialriptide** [comment]
> As MyreMylar expressed in the linked issue how this could be a new function, I quite like that suggestion myself too

I wouldn't mind this either.

**Comment by novialriptide** [comment]
A new PR is in the works right now, it'll either be `pygame.Surface.paste()` or something else... I'll close this PR for now as this will no longer be in use.

**Comment by novialriptide** [comment]
Reopening since it seems like `.blit()` is favored.

**Comment by MyreMylar** [comment]
I think sticking this in blit is fine if we are also having raw blit without all the bells and whistles. I don't think it will have that much performance impact anyway (and there is still performance work to be done on the slower alpha blits at the pixel level anyway)

**Comment by novialriptide** [comment]
> I think sticking this in blit is fine if we are also having raw blit without all the bells and whistles. I don't think it will have that much performance impact anyway (and there is still performance work to be done on the slower alpha blits at the pixel level anyway)

Got it, the performance loss shouldn't be too dramatically different anyway if there is any. (since it should be just an if statement or smth)

**Comment by novialriptide** [comment]
I realized that there was a big problem, so I did this little workaround to fix it.

Right now, we think that legacy code could work as such like this (let's say `mario_bulletbill.png` is an existing image):
```python
bulletbill = pygame.image.load("mario_bulletbill.png")
s = pygame.Surface((256, 224))
s.blit(bulletbill, (0, 0), area=None, special_flags=BLEND_SUB)
```
But the problem is that the kwarg values could be inputted like this
```python
bulletbill = pygame.image.load("mario_bulletbill.png")
s = pygame.Surface((256, 224))
s.blit(bulletbill, (0, 0), None, BLEND_SUB)
```

The code will detect `(0, 0)` as the `topleft` kwarg, `None` as the `topright` kwarg, and so on.

I don't think this should be too big of a problem anyways. I'm just pointing it out.

**Comment by novialriptide** [comment]
All tests should pass now.
image

**Comment by novialriptide** [comment]
## Performance Test
## New code (using this branch)
```python
import pygame
import timeit
surf = pygame.Surface((255, 255))
surf2 = pygame.Surface((10, 10))
surf2.fill((255, 0, 0))

timed = timeit.timeit(lambda: surf.blit(surf2, topright=(0, 0)), number=1)
print(timed, "ms")
```
```shell
pygame 2.1.3.dev5 (SDL 2.0.22, Python 3.9.0)
Hello from the pygame community. https://www.pygame.org/contribute.html
1.5399999999998748e-05 ms
```
## Old code (using this branch)
```python
import pygame
import timeit
surf = pygame.Surface((255, 255))
surf2 = pygame.Surface((10, 10))
surf2.fill((255, 0, 0))

timed = timeit.timeit(lambda: surf.blit(surf2, (0, 0)), number=1)
print(timed, "ms")
```
```shell
pygame 2.1.3.dev5 (SDL 2.0.22, Python 3.9.0)
Hello from the pygame community. https://www.pygame.org/contribute.html
1.580000000001025e-05 ms
```
## Old code (pygame 2.1.2)
```python
import pygame
import timeit
surf = pygame.Surface((255, 255))
surf2 = pygame.Surface((10, 10))
surf2.fill((255, 0, 0))

timed = timeit.timeit(lambda: surf.blit(surf2, (0, 0)), number=1)
print(timed, "ms")
```
```shell
pygame 2.1.2 (SDL 2.0.18, Python 3.9.0)
Hello from the pygame community. https://www.pygame.org/contribute.html
1.1000000000011001e-05 ms
```

**Comment by novialriptide** [comment]
Here's another performance test
## Performance Test (part 2)
### pygame 2.1.3.dev5 (this branch)
```python
import pygame
import timeit
surf = pygame.Surface((255, 255))
surf2 = pygame.Surface((10, 10))
surf2.fill((255, 0, 0))

timed = timeit.timeit(lambda: surf.blit(surf2, topleft=(100, 100)), number=10000000)
print(f"topleft: {timed} seconds")

timed = timeit.timeit(lambda: surf.blit(surf2, topright=(100, 100)), number=10000000)
print(f"topright: {timed} seconds")

timed = timeit.timeit(lambda: surf.blit(surf2, bottomleft=(100, 100)), number=10000000)
print(f"bottomleft: {timed} seconds")

timed = timeit.timeit(lambda: surf.blit(surf2, bottomright=(100, 100)), number=10000000)
print(f"bottomright: {timed} seconds")

timed = timeit.timeit(lambda: surf.blit(surf2, midleft=(100, 100)), number=10000000)
print(f"midleft: {timed} seconds")

timed = timeit.timeit(lambda: surf.blit(surf2, midright=(100, 100)), number=10000000)
print(f"midright: {timed} seconds")

timed = timeit.timeit(lambda: surf.blit(surf2, midtop=(100, 100)), number=10000000)
print(f"midtop: {timed} seconds")

timed = timeit.timeit(lambda: surf.blit(surf2, midbottom=(100, 100)), number=10000000)
print(f"midbottom: {timed} seconds")

timed = timeit.timeit(lambda: surf.blit(surf2, center=(100, 100)), number=10000000)
print(f"center: {timed} seconds")
```
topleft: 4.0847878 seconds
topright: 5.235740300000001 seconds
bottomleft: 5.8812159999999984 seconds
bottomright: 6.3172087999999995 seconds
midleft: 6.667041000000001 seconds
midright: 7.170022800000002 seconds
midtop: 7.438604399999996 seconds
midbottom: 7.826056099999995 seconds
center: 8.368447199999999 seconds
```

### pygame 2.1.2
```python
import pygame
import timeit
surf = pygame.Surface((255, 255))
surf2 = pygame.Surface((10, 10))
surf2.fill((255, 0, 0))

timed = timeit.timeit("surf.blit(surf2, surf2.get_rect(topleft=(0, 0)))", globals={"surf": surf, "surf2": surf2}, number=10000000)
print(f"topleft: {timed} seconds")

timed = timeit.timeit("surf.blit(surf2, surf2.get_rect(topright=(0, 0)))", globals={"surf": surf, "surf2": surf2}, number=10000000)
print(f"topright: {timed} seconds")

timed = timeit.timeit("surf.blit(surf2, surf2.get_rect(bottomleft=(0, 0)))", globals={"surf": surf, "surf2": surf2}, number=10000000)
print(f"bottomleft: {timed} seconds")

timed = timeit.timeit("surf.blit(surf2, surf2.get_rect(bottomright=(0, 0)))", globals={"surf": surf, "surf2": surf2}, number=10000000)
print(f"bottomright: {timed} seconds")

timed = timeit.timeit("surf.blit(surf2, surf2.get_rect(midleft=(0, 0)))", globals={"surf": surf, "surf2": surf2}, number=10000000)
print(f"midleft: {timed} seconds")

timed = timeit.timeit("surf.blit(surf2, surf2.get_rect(midright=(0, 0)))", globals={"surf": surf, "surf2": surf2}, number=10000000)
print(f"midright: {timed} seconds")

timed = timeit.timeit("surf.blit(surf2, surf2.get_rect(midtop=(0, 0)))", globals={"surf": surf, "surf2": surf2}, number=10000000)
print(f"midtop: {timed} seconds")

timed = timeit.timeit("surf.blit(surf2, surf2.get_rect(midbottom=(0, 0)))", globals={"surf": surf, "surf2": surf2}, number=10000000)
print(f"midbottom: {timed} seconds")

timed = timeit.timeit("surf.blit(surf2, surf2.get_rect(center=(0, 0)))", globals={"surf": surf, "surf2": surf2}, number=10000000)
print(f"center: {timed} seconds")
```

```shell
pygame 2.1.3.dev5 (SDL 2.0.22, Python 3.9.0)
Hello from the pygame community. https://www.pygame.org/contribute.html
topleft: 3.7194558 seconds
topright: 3.1512507000000003 seconds
bottomleft: 3.1497751000000003 seconds
bottomright: 3.168838599999999 seconds
midleft: 3.5747503 seconds
midright: 3.1675736000000008 seconds
midtop: 3.5943006000000004 seconds
midbottom: 3.190688999999999 seconds
center: 3.4581815999999996 seconds
```

## Conclusion
I need to optimize this a bit...

**Comment by novialriptide** [comment]
## Progress Update
I did some research progress today and it seems like the working following code is a lot slower than the cluttered bunch of if statements that we currently have right now. (Sorry for the noise)

Also, it seems that `PyArg_ParseTupleAndKeywords()` is actually the main cause of slow performance here since we have a lot of kwargs that the C code needs to process. If I switch `topleft` to the very end and `center` to the very start in `PyArg_ParseTupleAndKeywords()`, `center` gets a significant performance boost while `topleft` gets really slow.
```c
static PyObject *
surf_blit(pgSurfaceObject *self, PyObject *args, PyObject *keywds)
{
    SDL_Surface *src, *dest = pgSurface_AsSurface(self);
    SDL_Rect *src_rect, temp;
    PyObject *argrect = NULL;
    pgSurfaceObject *srcobject;
    int dx, dy, result;
    SDL_Rect dest_rect;
    int sx, sy;
    int the_args = 0; /* Represents kwarg: "special_flags". */

    int checks = 0;
    int loop;
    PyObject *origin_pos = NULL;

    /* 1 = 100% of the width/height, 2 = 50% of the width/height */
    PyObject *rect;

    PyObject *topleft_pos = NULL;
    PyObject *topright_pos = NULL;
    PyObject *bottomleft_pos = NULL;
    PyObject *bottomright_pos = NULL;
    PyObject *midleft_pos = NULL;
    PyObject *midright_pos = NULL;
    PyObject *midtop_pos = NULL;
    PyObject *midbottom_pos = NULL;
    PyObject *center_pos = NULL;
    PyObject *legacy_pos = NULL;

    static char *kwids[] = {"source",        "topleft",   "area",
                            "special_flags", "topright",  "bottomleft",
                            "bottomright",   "midleft",   "midright",
                            "midtop",        "midbottom", "center",
                            "dest",          NULL};
    if (!PyArg_ParseTupleAndKeywords(
            args, keywds, "O!|OOiOOOOOOOOO", kwids, &pgSurface_Type,
            &srcobject, &topleft_pos, &argrect, &the_args, &topright_pos,
            &bottomleft_pos, &bottomright_pos, &midleft_pos, &midright_pos,
            &midtop_pos, &midbottom_pos, &center_pos, &legacy_pos))
        return NULL;

    PyObject *pos_vars[] = {topleft_pos,     topright_pos,  bottomleft_pos,
                            bottomright_pos, midleft_pos,   midright_pos,
                            midtop_pos,      midbottom_pos, center_pos,
                            legacy_pos};

    for (loop = 0; loop < (int)(sizeof(pos_vars) / sizeof(pos_vars[0]));
         loop++) {
        if (pos_vars[loop] != NULL) {
            checks += 1;
        }
    }

    if (checks > 1) {
        return RAISE(PyExc_ValueError,
                     "More than one position kwarg identified.");
    }

    src = pgSurface_AsSurface(srcobject);

    if (!dest || !src)
        return RAISE(pgExc_SDLError, "display Surface quit");

    rect = pgRect_New4(0, 0, src->w, src->h);

    if (topleft_pos && (src_rect = pgRect_FromObject(topleft_pos, &temp))) {
        dx = src_rect->x;
        dy = src_rect->y;
    }
    else if (rect && keywds) {
        PyObject *key, *value;
        Py_ssize_t pos = 0;

        while (PyDict_Next(keywds, &pos, &key, &value)) {
            if ((PyObject_SetAttr(rect, key, value) == -1)) {
                Py_DECREF(rect);
                return NULL;
            }
        }
        dx = PyLong_AsLong(PyObject_GetAttrString(rect, "x"));
        dy = PyLong_AsLong(PyObject_GetAttrString(rect, "y"));
    }
    else
        return RAISE(PyExc_TypeError, "invalid destination position for blit");

    if (argrect && argrect != Py_None) {
        if (!(src_rect = pgRect_FromObject(argrect, &temp)))
            return RAISE(PyExc_TypeError, "Invalid rectstyle argument");
    }
    else {
        temp.x = temp.y = 0;
        temp.w = src->w;
        temp.h = src->h;
        src_rect = &temp;
    }

    if (!dest || !src)
        return RAISE(pgExc_SDLError, "display Surface quit");

    dest_rect.x = dx;
    dest_rect.y = dy;
    dest_rect.w = src->w;
    dest_rect.h = src->h;

    if (!the_args) {
        the_args = 0;
    }

    result = pgSurface_Blit(self, srcobject, &dest_rect, src_rect, the_args);

    if (result != 0) {
        return NULL;
    }

    return pgRect_New(&dest_rect);
}
```

### Performance results
```
topleft: 4.9320799 seconds
topright: 5.9666245 seconds
bottomleft: 6.4147993 seconds
bottomright: 6.806393800000002 seconds
midleft: 7.2275171 seconds
midright: 7.536997499999998 seconds
midtop: 7.813934099999997 seconds
midbottom: 8.3068691 seconds
center: 8.664110199999996 seconds
```

File: src_c/surface.c [hunk 2]
```diff
@@ -1824,31 +1824,62 @@ surf_blit(pgSurfaceObject *self, PyObject *args, PyObject *keywds)
...
+                          &topleft_pos))
         return NULL;
 
     src = pgSurface_AsSurface(srcobject);
+
     if (!dest || !src)
         return RAISE(pgExc_SDLError, "display Surface quit");
 
-    if ((src_rect = pgRect_FromObject(argpos, &temp))) {
+    rect = pgRect_New4(0, 0, src->w, src->h);
```
Comment (itzpr3d4t0r) [inline src_c/surface.c:1849]: We are creating this rect each call, but we only use it when we have kwargs, so better move it inside the `else if((rect && keywds)`, and also make change the if to `else if (keywds)`
```suggestion
```

File: src_c/surface.c [hunk 3]
```diff
@@ -1824,31 +1824,62 @@ surf_blit(pgSurfaceObject *self, PyObject *args, PyObject *keywds)
...
+    PyObject *argrect = NULL;
     pgSurfaceObject *srcobject;
     int dx, dy, result;
     SDL_Rect dest_rect;
     int sx, sy;
-    int the_args = 0;
+    int the_args = 0; /* Represents kwarg: "special_flags". */
+
+    int checks = 0;
+    int loop;
```
Comment (itzpr3d4t0r) [inline src_c/surface.c:1835]: These two are unused

File: src_c/surface.c [hunk 4]
```diff
@@ -1824,31 +1824,62 @@ surf_blit(pgSurfaceObject *self, PyObject *args, PyObject *keywds)
...
+                // if it's a number then it must be "special_flags"
+                if (!pg_IntFromObj(value, &the_args))
+                    return RAISE(PyExc_TypeError,
+                                 "special_flags parameter must be numeric");
+            }
+            else if (pgRect_Check(value)) {
+                argrect = value;
+            }
+        }
+        dx = PyLong_AsLong(PyObject_GetAttrString(rect, "x"));
```
Comment (itzpr3d4t0r) [inline src_c/surface.c:1880]: Isn't there a more efficient way of getting the `x` from a pygame Rect?

File: src_c/surface.c [hunk 5]
```diff
@@ -1863,18 +1894,23 @@ surf_blit(pgSurfaceObject *self, PyObject *args, PyObject *keywds)
         src_rect = &temp;
     }
 
+    if (!dest || !src)
+        return RAISE(pgExc_SDLError, "display Surface quit");
+
     dest_rect.x = dx;
     dest_rect.y = dy;
```
Comment (itzpr3d4t0r) [inline src_c/surface.c:1901]: I would delete these two and assign them wherever `dx` and `dy` are assigned
```suggestion
```

**Comment by novialriptide** [comment]
@Starbuck5 and I have decided that `anchor=` would be a better decision as the performance using rect-like kwargs are much worse than using `get_rect()`

**Pull request closed without merging**